pkg/providers/vsphere/vmlifecycle/update_status.go

```diff
@@ -678,3 +682,145 @@ func updateGuestNetworkStatus(vm *vmopv1.VirtualMachine, gi *vimtypes.GuestInfo)
+	if moVM.Summary.Storage == nil {
+		return
+	}
+	if vm.Status.Storage == nil {
+		vm.Status.Storage = &vmopv1.VirtualMachineStorageStatus{}
+	}
+
+	vm.Status.Storage.Committed = BytesToResourceGiB(
+		moVM.Summary.Storage.Committed)
+
+	vm.Status.Storage.Uncommitted = BytesToResourceGiB(
+		moVM.Summary.Storage.Uncommitted)
+
+	vm.Status.Storage.Unshared = BytesToResourceGiB(
+		moVM.Summary.Storage.Unshared)
+}
+
+func updateVolumeStatus(vm *vmopv1.VirtualMachine, moVM mo.VirtualMachine) {
+	if moVM.Config == nil ||
+		moVM.LayoutEx == nil ||
+		len(moVM.LayoutEx.Disk) == 0 ||
+		len(moVM.LayoutEx.File) == 0 ||
+		len(moVM.Config.Hardware.Device) == 0 {
+
+		return
+	}
+
+	existingDisksInStatus := map[string]int{}
+	for i := range vm.Status.Volumes {
+		existingDisksInStatus[vm.Status.Volumes[i].DiskUUID] = i
+	}
+
+	existingDisksInConfig := map[string]struct{}{}
+
+	for i := range moVM.Config.Hardware.Device {
+		vd, ok := moVM.Config.Hardware.Device[i].(*vimtypes.VirtualDisk)
+		if !ok {
+			continue
+		}
+
+		var (
+			diskUUID string
+			fileName string
+		)
+
+		switch tb := vd.Backing.(type) {
+		case *vimtypes.VirtualDiskFlatVer2BackingInfo:
+			diskUUID = tb.Uuid
+			fileName = tb.FileName
+		case *vimtypes.VirtualDiskSeSparseBackingInfo:
+			diskUUID = tb.Uuid
+			fileName = tb.FileName
+		case *vimtypes.VirtualDiskRawDiskMappingVer1BackingInfo:
+			diskUUID = tb.Uuid
+			fileName = tb.FileName
+		case *vimtypes.VirtualDiskSparseVer2BackingInfo:
+			diskUUID = tb.Uuid
+			fileName = tb.FileName
+		case *vimtypes.VirtualDiskRawDiskVer2BackingInfo:
+			diskUUID = tb.Uuid
+			fileName = tb.DescriptorFileName
+		}
+
+		var diskPath object.DatastorePath
+		if !diskPath.FromString(fileName) {
+			continue
+		}
+
+		existingDisksInConfig[diskUUID] = struct{}{}
+
+		ctx := context.TODO()
+
+		if diskIndex, ok := existingDisksInStatus[diskUUID]; ok {
+			// The disk is already in the list of volume statuses, so update the
+			// existing status with the usage information.
+			di, _ := vmdk.GetVirtualDiskInfoByUUID(ctx, nil, moVM, false, diskUUID)
+			vm.Status.Volumes[diskIndex].Used = BytesToResourceGiB(di.UniqueSize)
+		} else if !strings.HasPrefix(diskPath.Path, "fcd/") {
+			// The disk is a classic, non-FCD that must be added to the list of
+			// volume statuses.
+			di, _ := vmdk.GetVirtualDiskInfoByUUID(ctx, nil, moVM, false, diskUUID)
+			dp := diskPath.Path
+			vm.Status.Volumes = append(
+				vm.Status.Volumes,
+				vmopv1.VirtualMachineVolumeStatus{
+					Name:     strings.TrimSuffix(path.Base(dp), path.Ext(dp)),
+					Type:     vmopv1.VirtualMachineStorageDiskTypeClassic,
+					Attached: true,
+					DiskUUID: diskUUID,
+					Limit:    BytesToResourceGiB(di.CapacityInBytes),
+					Used:     BytesToResourceGiB(di.UniqueSize),
+				})
+		}
+	}
+
+	// Remove any status entries for classic disks that no longer exist in
+	// config.hardware.device.
+	i := 0
+	for _, vol := range vm.Status.Volumes {
```

Comment:
Nit - maybe https://pkg.go.dev/slices#DeleteFunc moderately easier to read?

Reply:
I forgot they added this, thank you.

Reply:
Actually, I am going to keep this current pattern, and you will see why in a moment.